Fix package definitions that include executable files

AltGr opened the pull request:
Future opam won't support file permissions within a repository (it makes
the abstraction much simpler for signing, etc., and really they
shouldn't be needed)

This PR fixes all packages that made use of the opam-repository stored
mode for files in `files/` subdirectories (shell scripts that were
called with `./script` from the opam file for almost all of them, which
just needed an explicit shell call).

Executable permission for all files has also been removed.

Note: there is no (lint) check for that at the moment.

camelus commented:
##### :x: opam-lint errors <small>c4ac1f1aacd2b0fc666d231cd0852c3844d59236</small>

* **`conf-libev.4-11`** has errors:

  * **error 24:** Field 'maintainer' has the old default value
  * **error 25:** Missing field 'authors'
  * **error 35:** Missing field 'homepage'
  * **warning 36:** Missing field 'bug-reports'
  * **warning 37:** Missing field 'dev-repo'



* **`gmp-xen.6.0.0`** has errors:

  * **error 25:** Missing field 'authors'
  * **warning 36:** Missing field 'bug-reports'
  * **warning 37:** Missing field 'dev-repo'

* **`gmp-freestanding.6.0.0`** has some warnings:

  * **warning 37**: Missing field 'dev-repo'



* **`sibylfs-lem.0.4.0`** has some warnings:

  * **warning 36**: Missing field 'bug-reports'
  * **warning 41**: Some packages are mentionned in package scripts of features, but there is no dependency or depopt toward them: "sibylfs-lem"



* **`zarith-freestanding.1.4`** has some warnings:

  * **warning 36**: Missing field 'bug-reports'
  * **warning 37**: Missing field 'dev-repo'



* **`zarith-xen.1.3`** has some warnings:

  * **warning 36**: Missing field 'bug-reports'
  * **warning 37**: Missing field 'dev-repo'



* **`zarith-xen.1.4`** has some warnings:

  * **warning 36**: Missing field 'bug-reports'
  * **warning 37**: Missing field 'dev-repo'

* These packages passed lint tests: `conf-pic-switch.0.1`, `llvm.3.4`, `llvm.3.5`, `llvm.3.6`, `llvm.3.7`, `llvm.3.8`, `llvm.3.9`, `minios-xen.0.9`, `ocamlfind.1.6.2`, `ocamlfind.1.7.1`, `ssl.0.4.6`, `ssl.0.4.6a`, `ssl.0.4.7`, `ssl.0.5.0`


---


##### :white_check_mark: Installability check (5800 &rarr; 5800)

AltGr commented:
> warning 41: Some packages are mentionned in package scripts of features, but there is no dependency or depopt toward them: "sibylfs-lem"

Hmm, I thought this had been fixed in 1.3 :/

AltGr commented:
Is this ok to merge ? The changes are only:
- removed `+x` permissions from all files
- in opam files, prefixed calls to `./script` when `script` was included in `files/` with the command found after the `#!` (all were shell calls) — so that should be functionally equivalent, but without requiring `+x`.
- ocamlfind installed an included shell script with `cp`, replaced by `install` so that `+x` is set when installing.

Travis fails on llvm 3.9 beacause of outdated system dependencies, after succeeding on many others (would be nice to continue and report what packages failed at the end ?). This is, in any case, unrelated to the PR.

samoht commented on packages/ssl/ssl.0.5.0/opam:
```diff
@@ -6,8 +6,8 @@ dev-repo: "https://github.com/savonet/ocaml-ssl.git"
-  ["./pkgconfigure" "--prefix" prefix] {os = "openbsd"}
-  ["./configure" "--prefix" prefix] {os != "openbsd"}
+  ["/bin/sh" "./pkgconfigure" "--prefix" prefix] {os = "openbsd"}
+  ["" "/bin/sh" "./configure" "--prefix" prefix] {os != "openbsd"}
```
why the extra `""` here?

AltGr replied:
oops, looks like a rewriting bug...

samoht commented:
Does hard-coding `/bin/sh` makes thing worst or better on windows? /cc @dra27 and @fdopen 

Also I see that @ygrek did something a bit similar in #7934  @ygrek would you mind commenting if that PR goes in the right direction?

avsm commented on packages/conf-pic-switch/conf-pic-switch.0.1/opam:
```diff
@@ -5,7 +5,7 @@ homepage: "https://github.com/ocaml/opam-repository"
 dev-repo: "https://github.com/ocaml/opam-repository.git"
 build: [
-  ["./check.sh"]
+  ["/bin/bash" "./check.sh"]
```
this is not a portable invocation of bash unfortunately -- it will be in /usr/local in BSDs. So just `bash` should be sufficient as it will be in the parent PATH

fdopen commented:
@ygrek uses ` [ "sh" ... ]` in https://github.com/ocaml/opam-repository/pull/7933 - for termux support. That would certainly be better for windows as well.

ygrek commented:
I am strongly in favor calling "sh" instead of "/bin/sh" - even POSIX doesn't guarantee the latter, see http://pubs.opengroup.org/onlinepubs/009695399/utilities/sh.html : 
> Applications should note that the standard PATH to the shell cannot be assumed to be either /bin/sh or /usr/bin/sh, and should be determined by interrogation of the PATH returned by getconf PATH , ensuring that the returned pathname is an absolute pathname and not a shell built-in.

Even better would be probably to have opam define a variable `sh = command -v sh` (like it does for make) and use that (and allow override in case user wants something strange).

AltGr commented:
Ok, I added `"sh"` initially, but then switched to recovering the strict shebangs from the script files (including e.g. `-eux` options) to remain closer to the original. But if we can take the chance to improve compatibility by removing `/bin/`, I'd seize the opportunity.

> Even better would be probably to have opam define a variable sh = command -v sh (like it does for make) and use that (and allow override in case user wants something strange).

Good suggestion, but that can't be changed easily in existing installations of opam 1.2. What I can do is add such a variable in opam 2, so that we become able to replace `"sh"` with `sh` in the future. (opam 2 allows easy definition of variables through `opam config set VAR VALUE`, and in either case, `make` is only specific in that it is automatically defined at switch creation)

avsm commented:
This lgtm. Would be good to smoketest it on FreeBSD. @hannesm, any views on this?

hannesm commented:
I don't have enough time and energy to properly test this on FreeBSD (esp since I don't care about all of the packages)... but I'm curious why sometimes you use `sh -eux`, sometimes no arguments, sometimes only a subset of `eux`....

should we really use `sh` and `bash`?  or rather `/usr/bin/env sh`/`/usr/bin/env bash`.  I'm all for some uniform way (and would appreciate to get rid of GNUisms like using bash scripts for no good reason).

ygrek commented:
/usr/bin/env is needed only in shebang, as a way to lookup binary in PATH w/o specifying absolute path. When called from shell we can rely on shell doing PATH lookup so no need for any env helpers, plain "sh" or "bash" is enough. As for the flags, I think it is better to `set` them inside script explicitly, than rely on the caller to supply them (because they change behaviour significantly), but it is up to script's authors, not opam's responsibility to change (same as bash vs sh).

avsm commented:
thank you for the comprehensive patch @AltGr!
